### PR TITLE
Closes #8784 Create download directory when it doesn't exists

### LINKED ISSUE #8784: Create download directory when it doesn't exists
Related https://github.com/mozilla-mobile/fenix/issues/5829

### PR DESCRIPTION
**Pull request by Amejia481**
For legacy file API we need to make sure that the directory where we want to write a download exists, for the new api (scoped storage) this is not the case.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ permalink: /changelog/
 * **concept-engine**
   * ⚠️ Removed `TrackingCategory`.`SHIMMED`, for user usability reasons, we are going to mark SHIMMED categories as blocked, to follow the same pattern as Firefox desktop for more information see [#8769](https://github.com/mozilla-mobile/android-components/issues/8769)
 
+* **feature-downloads**
+  * 🚒 Bug fixed [issue #8585](https://github.com/mozilla-mobile/android-components/issues/8784) create download directory when it doesn't exists for more information see [mozilla-mobile/fenix#15527](https://github.com/mozilla-mobile/fenix/issues/15527).
```
Comment (csadilek) [inline docs/changelog.md:21]: This links to the wrong Fenix issue.

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt [hunk 2]
```diff
@@ -825,9 +825,18 @@ abstract class AbstractFetchDownloadService : Service() {
     @Suppress("Deprecation")
     @VisibleForTesting
     internal fun useFileStreamLegacy(download: DownloadState, append: Boolean, block: (OutputStream) -> Unit) {
+        createDirectoryIfNoExists(download)
         FileOutputStream(File(download.filePath), append).use(block)
     }
 
+    @VisibleForTesting
+    internal fun createDirectoryIfNoExists(download: DownloadState) {
```
Comment (csadilek) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:833]: nit: if not exists, but better `createDirectoryIfNeeded` ?

File: components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt [hunk 3]
```diff
@@ -1463,6 +1464,20 @@ class AbstractFetchDownloadServiceTest {
         assertEquals(2, shadowNotificationService.size())
     }
 
+    @Test
+    fun `createDirectoryIfNoExists - MUST create directory when it does not exists`() = runBlocking {
```
Comment (csadilek) [inline components/feature/downloads/src/test/java/mozilla/components/feature/downloads/AbstractFetchDownloadServiceTest.kt:1468]: nit: name see above.

**Review by csadilek — APPROVED** [review]
✅ Changelog needs an update and I think we should rename the method.

File: docs/changelog.md (hunk 1 above)
Comment (Amejia481) [reply docs/changelog.md:21]: Fixing, thanks!

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt (hunk 2 above)
Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:833]: 👍🏽

**Pull request merged**